### PR TITLE
Remove scheduled test

### PR DESCRIPTION
**Pull request by grouma**
- Remove dep on `scheduled_test`
- Depend on latest `package:test`
- Remove duplicate `pumpEventQueue`
- Use `pumpEventQueue` instead of `startClosingEventStream`
- Use `test_descriptor` instead of manually creating sandbox
- Remove unnecessary `Backup case` and consequently `allowEither`.

Manually verified on Linux, Mac and Windows.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# 0.9.8
```
Comment (nex3) [inline CHANGELOG.md]: This should be 0.9.7+6. There aren't any new user-facing features here.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
...
+
+* Remove dep on scheduled test.
```
Comment (nex3) [inline CHANGELOG.md]: I'd just write "Internal changes only" so that users don't have to spend mental energy thinking about whether this release is relevant to them.

File: pubspec.yaml [hunk 3]
```diff
@@ -12,5 +12,5 @@ dependencies:
   path: '>=0.9.0 <2.0.0'
 dev_dependencies:
   benchmark_harness: '^1.0.4'
-  scheduled_test: '^0.12.0'
-  test: '^0.12.18+1'
+  test: '^0.12.29+1'
```
Comment (nex3) [inline pubspec.yaml]: Why `+1`? Lower bounds should generally be pinned to versions that introduce new features.

File: pubspec.yaml [hunk 4]
```diff
@@ -12,5 +12,5 @@ dependencies:
...
+  test_descriptor: '^1.0.3'
```
Comment (nex3) [inline pubspec.yaml]: Same question here—why not `^1.0.0`?

File: test/directory_watcher/mac_os_test.dart [hunk 5]
```diff
@@ -25,36 +23,31 @@ void main() {
...
+        (i, j, k) => isRemoveEvent("dir/sub/sub-$i/sub-$j/file-$k.txt")));
 
-      inAnyOrder(withPermutations(
-          (i, j, k) => isAddEvent("dir/sub/sub-$i/sub-$j/file-$k.txt")));
-    }, () {
-      inAnyOrder(withPermutations(
-          (i, j, k) => isModifyEvent("dir/sub/sub-$i/sub-$j/file-$k.txt")));
-    });
+    await inAnyOrder(withPermutations(
+        (i, j, k) => isAddEvent("dir/sub/sub-$i/sub-$j/file-$k.txt")));
```
Comment (nex3) [inline test/directory_watcher/mac_os_test.dart]: Why isn't this using `allowEither()` anymore?

File: test/directory_watcher/shared.dart [hunk 6]
```diff
@@ -2,270 +2,241 @@
...
     });
   });
 
-  // Most of the time, when multiple filesystem actions happen in sequence,
-  // they'll be batched together and the watcher will see them all at once.
-  // These tests verify that the watcher normalizes and combine these events
-  // properly. However, very occasionally the events will be reported in
-  // separate batches, and the watcher will report them as though they occurred
-  // far apart in time, so each of these tests has a "backup case" to allow for
-  // that as well.
```
Comment (nex3) [inline test/directory_watcher/shared.dart]: Why aren't we doing this anymore?

File: test/directory_watcher/windows_test.dart [hunk 7]
```diff
@@ -4,19 +4,19 @@
...
 
 void main() {
   watcherFactory = (dir) => new WindowsDirectoryWatcher(dir);
 
-  setUp(createSandbox);
-
-  sharedTests();
+  // TODO(grouma) - renable when https://github.com/dart-lang/sdk/issues/31760
+  // is resolved.
+  // sharedTests();
```
Comment (nex3) [inline test/directory_watcher/windows_test.dart]: Skip tests rather than commenting them out.

File: test/no_subscription/shared.dart [hunk 8]
```diff
@@ -4,60 +4,59 @@
...
     // Subscribe to the events.
     var completer = new Completer();
-    var subscription = watcher.events.listen(expectAsync1((event) {
-      expect(event, isWatchEvent(ChangeType.ADD, "file.txt"));
+    queue.next.then((event) {
+      expect(event.type, ChangeType.ADD);
+      expect(event.path.contains("file.txt"), isTrue);
       completer.complete();
-    }));
+    });
```
Comment (nex3) [inline test/no_subscription/shared.dart]: I think you can express this more cleanly as:

```dart
var future = expectLater(queue, emits(isWatchEvent(ChangeType.ADD, "file.txt")));
expect(queue, emitsNothing);
```

The `emitsNothing` assertion does the same thing as `expectAsync1()` in the old code by ensuring no unexpected events are emitted.

Same below.

File: test/ready/shared.dart [hunk 9]
```diff
@@ -2,97 +2,60 @@
...
 import '../utils.dart';
 
 void sharedTests() {
-  test('ready does not complete until after subscription', () {
-    var watcher = createWatcher(waitForReady: false);
-
-    var ready = false;
-    watcher.ready.then((_) {
-      ready = true;
-    });
```
Comment (nex3) [inline test/ready/shared.dart]: It's important to check when the future fired—we're testing that it fires at the correct time, and we shouldn't rely on it matching `watcher.isReady`. Same below.

File: test/ready/shared.dart [hunk 10]
```diff
@@ -2,97 +2,60 @@
...
-      ready = false;
-      watcher.ready.then((_) {
-        ready = true;
-      });
+    await subscription.cancel();
+
+    // Track when it's ready again.
+    ready = false;
+    watcher.ready.then((_) {
+      ready = true;
```
Comment (nex3) [inline test/ready/shared.dart]: You can simplify this a lot by writing `expect(watcher.ready, doesNotComplete)`.

File: test/utils.dart [hunk 11]
```diff
@@ -2,21 +2,22 @@
...
 import 'dart:io';
 
+import 'package:async/async.dart';
 import 'package:path/path.dart' as p;
-import 'package:scheduled_test/scheduled_stream.dart';
-import 'package:scheduled_test/scheduled_test.dart';
-import 'package:watcher/watcher.dart';
+import 'package:test/test.dart';
+import 'package:test_descriptor/test_descriptor.dart' as d;
 import 'package:watcher/src/stat.dart';
```
Comment (nex3) [inline test/utils.dart]: Nit: add a blank line between imports of external packages and imports of the package being tested.

File: test/utils.dart [hunk 12]
```diff
@@ -27,129 +28,51 @@ Watcher _watcher;
...
-void startClosingEventStream() {
-  schedule(() {
-    _closePending = true;
-    pumpEventQueue().then((_) => _watcherEvents.close()).whenComplete(() {
-      _closePending = false;
-    });
-  }, 'start closing event stream');
+    var mtime = _mockFileModificationTimes[path];
+    return new DateTime.fromMillisecondsSinceEpoch(mtime == null ? 0 : mtime);
+  });
```
Comment (nex3) [inline test/utils.dart]: Nit: add a newline after this.

File: test/utils.dart [hunk 13]
```diff
@@ -27,129 +28,51 @@ Watcher _watcher;
...
-/// normally only closed after the test completes.
-void startClosingEventStream() {
-  schedule(() {
-    _closePending = true;
-    pumpEventQueue().then((_) => _watcherEvents.close()).whenComplete(() {
-      _closePending = false;
-    });
-  }, 'start closing event stream');
+    var mtime = _mockFileModificationTimes[path];
+    return new DateTime.fromMillisecondsSinceEpoch(mtime == null ? 0 : mtime);
```
Comment (nex3) [inline test/utils.dart]: Nit: `??`

File: test/utils.dart [hunk 14]
```diff
@@ -27,129 +28,51 @@ Watcher _watcher;
...
+  });
+  // We want to wait until we're ready *after* we subscribe to the watcher's
+  // events.
+  var watcher = createWatcher(path: path);
+  _watcherEvents = new StreamQueue(watcher.events);
+  // Forces a subscription to the underlying stream
```
Comment (nex3) [inline test/utils.dart]: Nit: punctuation

File: test/utils.dart [hunk 15]
```diff
@@ -175,47 +98,42 @@ StreamMatcher _collectStreamMatcher(block()) {
...
+    _expectOrCollect(mayEmit(_collectStreamMatcher(block)));
 
 /// Returns a matcher that matches a [WatchEvent] with the given [type] and
 /// [path].
-Matcher isWatchEvent(ChangeType type, String path) {
-  return predicate((e) {
-    return e is WatchEvent &&
-        e.type == type &&
-        e.path == p.join(_sandboxDir, p.normalize(path));
+StreamMatcher isWatchEvent(ChangeType type, String path) {
```
Comment (nex3) [inline test/utils.dart]: Why are you changing this from a `Matcher` to a `StreamMatcher`?

File: test/utils.dart [hunk 16]
```diff
@@ -175,47 +98,42 @@ StreamMatcher _collectStreamMatcher(block()) {
...
+  return new StreamMatcher((queue) async {
```
Comment (nex3) [inline test/utils.dart]: This doesn't follow most of the requirements in the [`new StreamMatcher()` documentation](https://www.dartdocs.org/documentation/test/0.12.30/test/StreamMatcher/StreamMatcher.html).

File: test/utils.dart [hunk 17]
```diff
@@ -27,129 +28,51 @@ Watcher _watcher;
...
+    // Make sure we got a path in the sandbox.
+    assert(p.isRelative(path) && !path.startsWith(".."));
 
-/// Schedule closing the watcher stream after the event queue has been pumped.
-///
-/// This is necessary when events are allowed to occur, but don't have to occur,
-/// at the end of a test. Otherwise, if they don't occur, the test will wait
-/// indefinitely because they might in the future and because the watcher is
-/// normally only closed after the test completes.
-void startClosingEventStream() {
```
Comment (nex3) [inline test/utils.dart]: Why was this deleted?

File: test/utils.dart (hunk 16 above)
Comment (grouma) [reply test/utils.dart]: I opted to go with no description because the printed details are already clear. I have updated to ensure it doesn't consume a value if it isn't valid.

File: test/directory_watcher/shared.dart (hunk 6 above)
Comment (grouma) [reply test/directory_watcher/shared.dart]: This doesn't seem to be an issue. I have yet to run into the 'backup case'. Also with the new test model, I'm not sure how I would even implement backup expects.

File: CHANGELOG.md (hunk 2 above)
Comment (nex3) [reply CHANGELOG.md]: The changelog is an aspect of API documentation—it should be targeted at downstream users of the package. Details like this that are only relevant to package maintainers belong in the commit log.

File: pubspec.yaml (hunk 3 above)
Comment (nex3) [reply pubspec.yaml]: Pinning to the latest version is always *safe*, but it's not usually the best thing to do as a responsible member of the package ecosystem. You want to make the version range as wide as possible while guaranteeing that the features your package uses are available to minimize version churn and the possibility of version lock.

File: test/directory_watcher/mac_os_test.dart (hunk 5 above)
Comment (nex3) [reply test/directory_watcher/mac_os_test.dart]: Just because it does so in your local tests doesn't mean it *always* does so. These alternate cases came about through painstaking debugging of intermittent edge cases that crop up in specific contexts. It's not safe to just delete them because the tests pass locally.

File: test/directory_watcher/shared.dart (hunk 6 above)
Comment (nex3) [reply test/directory_watcher/shared.dart]: See above—backup cases are important, and you can't reliably tell whether they're relevant with local testing. These systems are *extremely* contingent on opaque OS specifics, which is why the tests have so many alternate cases in the first place.

File: test/utils.dart (hunk 17 above)
Comment (nex3) [reply test/utils.dart]: This smells funny to me... why don't we need to cancel the stream anymore?

File: test/utils.dart (hunk 15 above)
Comment (nex3) [reply test/utils.dart]: I don't see the connection... `ScheduledStream` also had `StreamMatcher`s, and this wasn't one then. Why won't the old predicate matcher work?

File: test/utils.dart (hunk 16 above)
Comment (nex3) [reply test/utils.dart]: The description also needs to be in the past tense.

> I opted to go with no description because the printed details are already clear.

I don't like this. Not having an error message means this won't compose well with other matchers.

File: test/directory_watcher/windows_test.dart (hunk 7 above)
Comment (nex3) [reply test/directory_watcher/windows_test.dart]: It would probably be cleaner to add the group here, rather than manually forwarding parameters to `sharedTests()` and adding indentation and prefixes to every test (which also makes the diff a lot less clear).

File: test/directory_watcher/windows_test.dart [hunk 18]
```diff
@@ -8,15 +8,15 @@ import 'package:test/test.dart';
...
+import 'shared.dart';
 import '../utils.dart';
 
 void main() {
   watcherFactory = (dir) => new WindowsDirectoryWatcher(dir);
 
   // TODO(grouma) - renable when https://github.com/dart-lang/sdk/issues/31760
   // is resolved.
-  // sharedTests();
+  sharedTests(skipped: true);
```
Comment (nex3) [inline test/directory_watcher/windows_test.dart]: The skip should include a message describing why the tests are skipped.

File: test/ready/shared.dart (hunk 9 above)
Comment (nex3) [reply test/ready/shared.dart]: You still need to test *when* `watcher.ready` completes—it shouldn't be complete before `watcher.events` has a listener, and it should be complete after `watcher.ready` fires.

File: test/ready/shared.dart [hunk 19]
```diff
@@ -28,8 +28,14 @@ void sharedTests() {
...
 
-    // Should eventually be ready.
-    await watcher.ready;
+    // Allow watcher to become ready
+    await pumpEventQueue();
+
+    // Ensure ready completes immediately
+    expect(
+        watcher.ready.timeout(new Duration(milliseconds: 0),
+            onTimeout: () => throw 'Does not complete immedately'),
```
Comment (nex3) [inline test/ready/shared.dart]: This is a race condition. A watcher becoming ready involves OS interaction—there's no guarantee that a `pumpEventQueue()` call will last long enough for it to happen.

This test case should be essentially the same as it was in the `scheduled_test` world, with a `ready` variable and everything.

File: test/utils.dart (hunk 12 above)
Comment (nex3) [reply test/utils.dart]: I think you missed this comment.

File: test/ready/shared.dart (hunk 9 above)
Comment (grouma) [reply test/ready/shared.dart]: Should be good now.

File: test/utils.dart (hunk 17 above)
Comment (grouma) [reply test/utils.dart]: They weren't necessary when I removed the Backup cases. Now that these are present this function is now necessary.

File: test/directory_watcher/shared.dart (hunk 6 above)
Comment (grouma) [reply test/directory_watcher/shared.dart]: Added back the backup cases.

File: test/directory_watcher/mac_os_test.dart (hunk 5 above)
Comment (grouma) [reply test/directory_watcher/mac_os_test.dart]: Added back the backup cases.

**Comment by grouma** [comment]
Should be good now. PTAL.

File: test/directory_watcher/mac_os_test.dart (hunk 5 above)
Comment (nex3) [reply test/directory_watcher/mac_os_test.dart]: It looks like this backup case is still missing.

File: test/ready/shared.dart [hunk 20]
```diff
@@ -2,97 +2,61 @@
...
     // Subscribe to the events.
-    schedule(() {
-      var subscription = watcher.events.listen((event) {});
+    watcher.events.listen((event) {});
 
-      currentSchedule.onComplete.schedule(() {
-        subscription.cancel();
-      });
-    });
+    await expectLater(watcher.ready, completes);
```
Comment (nex3) [inline test/ready/shared.dart]: This can just be `await watcher.ready`. The `expectLater()` isn't doing anything here, since awaiting the future already guarantees the test won't complete until the future does.

**Comment by nex3** [comment]
FYI, I just published a version of `async` with a fix for the `StreamQueue` bug.

File: test/directory_watcher/mac_os_test.dart (hunk 5 above)
Comment (grouma) [reply test/directory_watcher/mac_os_test.dart]: Fixed.
